### PR TITLE
perf(django): avoid use of parse for URL parsing

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

The utility in the parse module are generic and generally expensive to
call. Its use in the Django integration has been removed in favour of
cheaper string concatenations.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

### CODE REVIEW

File: ddtrace/contrib/django/utils.py [hunk 1]
```diff
@@ -121,16 +120,9 @@ def get_request_uri(request):
...
-    # byte strings.
-    # https://github.com/python/cpython/blob/02d126aa09d96d03dcf9c5b51c858ce5ef386601/Lib/urllib/parse.py#L111-L125
-    if PY3 and not all(isinstance(value, binary_type) or value is None for value in urlparts.values()):
+    if PY3:
         for (key, value) in urlparts.items():
             if value is not None and isinstance(value, binary_type):
                 urlparts[key] = to_unicode(value)
 
-    return parse.urlunparse(parse.ParseResult(**urlparts))
+    return "".join((urlparts["scheme"], "://", urlparts["netloc"], urlparts["path"]))
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/django/utils.py:128]: I'm wary of moving away from the library function (but super happy to see the above stuff go away). What are the performance differences we're talking about here?

Comment (P403n1x87) [reply ddtrace/contrib/django/utils.py:128]: `get_request_uri` is about 9% of the total overhead. This change reduces that down to 3.5%

File: ddtrace/contrib/django/utils.py [hunk 2]
```diff
@@ -121,16 +120,9 @@ def get_request_uri(request):
...
 
-    # DEV: With PY3 urlunparse calls urllib.parse._coerce_args which uses the
-    # type of the scheme to check the type to expect from all url parts, raising
-    # a TypeError otherwise. If the scheme is not a str, the function returns
-    # the url parts bytes decoded along with a function to encode the result of
-    # combining the url parts. We returns a byte string when all url parts are
-    # byte strings.
-    # https://github.com/python/cpython/blob/02d126aa09d96d03dcf9c5b51c858ce5ef386601/Lib/urllib/parse.py#L111-L125
-    if PY3 and not all(isinstance(value, binary_type) or value is None for value in urlparts.values()):
+    if PY3:
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/django/utils.py:123]: @majorgreys do we need this logic anymore?

**Review by Kyle-Verhoog — APPROVED** [review]
lgtm

File: ddtrace/contrib/django/utils.py [hunk 3]
```diff
@@ -100,7 +99,7 @@ def get_request_uri(request):
 
     # Build request url from the information available
     # DEV: We are explicitly omitting query strings since they may contain sensitive information
-    urlparts = dict(scheme=request.scheme, netloc=host, path=request.path, params=None, query=None, fragment=None)
+    urlparts = dict(scheme=request.scheme, netloc=host, path=request.path)
```
Comment (Yun-Kim) [inline ddtrace/contrib/django/utils.py:102]:
```suggestion
    urlparts = {"scheme": request.scheme, "netloc": host, "path": request.path}
```
IIRC defining literal dictionaries are better performance-wise than `dict()`, might be worth changing this but not a concern

**Review by Yun-Kim — APPROVED** [review]
One small nit but LGTM otherwise

File: ddtrace/contrib/django/utils.py (hunk 2 above)
Comment (majorgreys) [reply ddtrace/contrib/django/utils.py:123]: We can coerce these values as we do elsewhere by using `six.ensure_text`, which helps us avoid having to handle the version differences.

File: ddtrace/contrib/django/utils.py (hunk 3 above)
Comment (Kyle-Verhoog) [reply ddtrace/contrib/django/utils.py:102]:
```
In [1]: %timeit dict(one="two")                                                                                                   
163 ns ± 1.39 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [2]: %timeit {"one": "two"}                                                                                                    
41.7 ns ± 0.317 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

**Comment by P403n1x87** [comment]
I'll follow up on this as there is some more clean-up that can be done

File: ddtrace/contrib/django/utils.py (hunk 3 above)
Comment (P403n1x87) [reply ddtrace/contrib/django/utils.py:102]: 👍 Yep, that's an excellent point. `dict` adds function calls so making an explicit dictionary is generally faster.

**Pull request merged**